lib/msf/core/db_manager/http/servlet/credential_servlet.rb

```diff
@@ -15,6 +15,7 @@ def self.registered(app)
 
   def self.get_credentials
     lambda {
+      warden.authenticate!
```

Comment:
It would be nice if we didn't have to add this in to every call definition. One method that might work is adding a `before` block to the `registered` method that calls it based on that specific endpoint.

Example:
`
def self.registered(app)
    app.before do
      warden.authenticate! if request.url =~ HostServlet.api_path
    end
    app.get HostServlet.api_path_with_id, &get_host
    app.post HostServlet.api_path, &report_host
    app.put HostServlet.api_path_with_id, &update_host
    app.delete HostServlet.api_path, &delete_host
    app.post HostServlet.api_search_path, &search
  end
`

Not sure if that exact code will work, but something along those lines.

Also, we should create a helper function for this somewhere so that way we can easily change things across all servlets if needed.

Reply:
We spoke about this via chat, but I wanted to document the key points in the PR. I started experimenting with this suggestion to remove the `warden.authenticate!` call from the individual lambda handlers the other day. There are a few issues: 1. the `before` block order of execution is not easily controlled and thus the authentication initialization check that occurs in the `before` block within `MetasploitApiApp` might not be executed prior to a servlet authenticate call resulting in unexpected behavior. 2. The `before` block from all of the servlets will run to check if the URL matches in order to perform the `authenticate!` call. For example, if a request for a `HostServlet` endpoint is made all of the other servlets `before` blocks might execute to check if the URL matches before reaching the `HostServlet` `before` block. 3. The actual URL match in the `before` block would have to be more complex to avoid future conflict, for example, `/api/v1/hosts` and `/api/v1/hosts-something-additional`. In addition, the logic may need to be more complex if a servlet has mixed scope or a mix of authenticated and unauthenticated endpoints such as `AuthServlet`.